src/travel-details-screens/utils.ts

```diff
@@ -179,3 +181,112 @@ function someLegsAreByTrain(tripPattern: TripPattern): boolean {
 export function isLegFlexibleTransport(leg: Leg): boolean {
   return !!leg.line?.flexibleLineType;
 }
+
+export function getPublicCodeFromLeg(leg: Leg): string {
+  return leg.fromPlace?.quay?.publicCode || leg.line?.publicCode || '';
+}
+
+function getLatestBookingDate(
+  latestBookingTime: string, // e.g. '15:16:00'
+  expectedStartTime: string, // e.g. '2023-07-14T17:56:32+02:00'
+): Date {
+  const expectedStartDate = new Date(expectedStartTime);
+  const latestBookingDate = dateWithReplacedTime(
+    expectedStartDate,
+    latestBookingTime,
+    'HH:mm:ss',
+  );
+
+  if (latestBookingDate.getTime() > expectedStartDate.getTime()) {
+    latestBookingDate.setDate(latestBookingDate.getDate() - 1);
+  }
+
+  return latestBookingDate;
+}
+
+const defaultBookingRequirement: BookingRequirement = {
+  bookingIsAvailable: false,
+  requiresBooking: false,
+  requiresBookingUrgently: false,
+  bookingAvailableImminently: false,
+  isTooLate: false,
+  isTooEarly: false,
+  secondsRemainingToDeadline: Infinity,
+  secondsRemainingToAvailable: Infinity,
+  latestBookingDate: new Date(Number.MAX_VALUE),
+  earliestBookingDate: new Date(Number.MAX_VALUE),
+};
+
+export function getBookingRequirementForLeg(
+  leg: Leg | undefined,
+  now: number,
+  flex_booking_number_of_days_available: number, // would be better to get earliestBookingTime as ISO string from Entur's API.
+): BookingRequirement {
```

Comment:
I would prefer not mapping this in the `Results` component, creating a new domain model for `TripPattern` (`AvailableTripPattern`) which is passed through the app. Generally we use the domain models from Entur/BFF throughout the app, and it is the business-specific lead nodes that utilise the data as needed.

This means for example to have a util function `isNecessaryBookingToEarly` which may be used in the leaf nodes, rather than passing around the value `isTooEarly` in a custom domain model. I feel this create less coupling between the different components in the app.

Reply:
Updated now as requested.
Even though the implementation was more concise originally, I can see why you would want this pattern implemented.